### DISCUSSION TITLE
Coupling MOOSE with external programm

### DISCUSSION THREAD

**Discussion by Fruitwind** [idaholab/moose#23636] [Q&A General] [answered]
Hello!
I am a new user of Moose and appreciate its capabilities, and would like to extend its application.
I would like to tightly couple Moose with the PhreeqcRm (an open-source, well-documented, and validated software for extensive geochemical modeling). The PhreeqcRM was developed specifically for coupling it with external transport codes and is supposed to be straightforward to connect.
 Whereas I found the "MOOSE-wrapped Apps" page interesting, I struggle with the lack of simple how-to instructions. I see the tight coupling as exchanging selected variables between MOOSE and Phreeqc on every time step for each node while performing necessary syntax transformation and processing via a third-party (MOOSE-based) app. 

So, my question is: how to do it? Is there a way to export variables to the external code, or is there another (maybe more elegant) way to couple two different syntax apps? I am apologizing if this is a too simple question, I am very new to the MOOSE and to the software development, but very enthusiastic that this coupling would be beneficial for the community.
Thanks,
Elena

**Comment by GiudGiud** [comment, marked as answer]
Hello

So if you want to call PhreeqcRM from MOOSE, you found the right resource. There's an example of doing this here:
https://github.com/neams-th-coe/cardinal
To do this, you will need to:
- compile both codes together (there'll be some basic Makefile work to do this)
- write some code that transfers a field (variable) from an application to another (examples in cardinal)

Note: tight coupling is, in our jargon, exchanging fields and repeating the time step until convergence. Exchanging once per time step is "loose coupling" in our jargon. 

There are alternatives to our system that are more or less resilient / long term:
- preCICE https://precice.org/, used by the UKAEA with MOOSE (https://github.com/aurora-multiphysics, you ll want to contact them to get the code, I dont see it rn)
- file based coupling. Just output the fields to a format the other code can run, and read them from the other code, alternate running each code with a script. This isnt as clean but it s easier to set up especially if you are running this on a regular computer. 

Guillaume

**Reply by Fruitwind** [reply]
Hi!
Thank you very much for your answers, I now take Cardinal as an example and try to learn from it, and will gladly use the chance to continue writing here :)
So essentially what I'd love to have is calling Phreeqc from MOOSE copying the parameter matrix for each node, and updating the field -sending it to MOOSE. At this stage, I assume that Phreeqc inputs at each node converge like a flash and no iterations are needed, which is probably a quite strong assumption. So this would be a "loose" coupling then. At the same time, I do not want to build new input files for Phreeqc on every time step, but just want to update variables values - and somehow I thought this will shorten the computational time and thus will be called "tight" coupling.. Sorry for the misuse of this term.

**Reply by GiudGiud** [reply]
No worries. That jargon is not universal anyway, just local to the MOOSE ecosystem. 

Ok please let us know how this goes!
